### ISSUE TITLE
incorrect padding error from addwarning

### ISSUE THREAD

**Issue by mapellidario** [dmwm/CRABServer#7089]
When we ban a user and forcefully fail a task, TW adds a warning to the task

https://github.com/dmwm/CRABServer/blob/82e032dcebd3f68b4f971d8349dd2b722b92e3cb/src/python/TaskWorker/MasterWorker.py#L362

But we had a situation where this request fails with error 500 [1].
 crabserver should not fail so badly if the inputs are malformed. Maybe we can be more careful with inputs validation. list of the failures at [2]
In this case it was particularly annoying becasue there were O(100) tasks to fail and so all TW slaves got stuck in longish sleep/retry look. See list of the failures at [3]

Stefano temporarily hid the problem under the carpet while we look for a proper solution [4].

### how to replicate

Not easy, but we can gather some information to better investigate the problem

versions:

- TW: `py3.220203`
- REST: `py3.220218`

tasks: 

- all of the form `20220222.*jmsilva_crab_SPRACE-BR-tests-.*`, such as `220222_143422:jmsilva_crab_SPRACE-BR-tests-1645540093-Analysis_FileBased-3`

these are the related log entries:

REST, [2]

```
> grep "binascii.Error: Incorrect padding" *startAt=18:00*
cmsweb-services-logs-zvwnq-startAt=18:00_0.txt:[22/Feb/2022:18:00:43]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-5t8s2
cmsweb-services-logs-sxbch-startAt=18:00_0.txt:[22/Feb/2022:18:00:53]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-t6fw6
cmsweb-services-logs-ph457-startAt=18:00_0.txt:[22/Feb/2022:18:01:32]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-srkw9
cmsweb-services-logs-zvwnq-startAt=18:00_1.txt:[22/Feb/2022:18:02:27]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-sscnr
cmsweb-services-logs-ph457-startAt=18:00_0.txt:[22/Feb/2022:18:03:56]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-srkw9
cmsweb-services-logs-x8j57-startAt=18:00_0.txt:[22/Feb/2022:18:05:35]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-ktnww
cmsweb-services-logs-x8j57-startAt=18:00_0.txt:[22/Feb/2022:18:07:38]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-ktnww
cmsweb-services-logs-ph457-startAt=18:00_1.txt:[22/Feb/2022:18:10:03]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-g4kp8
cmsweb-services-logs-5kg8b-startAt=18:00_0.txt:[22/Feb/2022:18:12:49]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-sn8bb
```

TW, [3]

```plaintext
> grep -A20 "2022-02-22 18:00:43,323" /data/container/TaskWorker/logs/twlog.txt
2022-02-22 18:00:43,323:DEBUG:MasterWorker,353:Forcefully failing task 220222_143422:jmsilva_crab_SPRACE-BR-tests-1645540093-Analysis_FileBased-3
2022-02-22 18:00:43,433:DEBUG:RESTInteractions,169:Sleeping 10 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:00:43 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': 'e9ef8f26d3c5a5b4b3b1027817bda8ed', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '15712.023 us', 'CMS-Server-Time': 'D=25656 t=1645549243405385', 'Connection': 'close'}
2022-02-22 18:00:53,529:DEBUG:RESTInteractions,169:Sleeping 39 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:00:53 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '7312a2e9ab7d30f0eeb6c927af802641', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '5511.045 us', 'CMS-Server-Time': 'D=63465 t=1645549253465316', 'Connection': 'close'}
2022-02-22 18:01:32,613:DEBUG:RESTInteractions,169:Sleeping 55 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:01:32 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '218457d73657d25d1ef99c048c664e0c', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '6405.354 us', 'CMS-Server-Time': 'D=15822 t=1645549292594348', 'Connection': 'close'}
2022-02-22 18:02:27,721:DEBUG:RESTInteractions,169:Sleeping 89 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:02:27 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '429bfb597f1fec6cb00f06e61ed2ef53', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '4264.355 us', 'CMS-Server-Time': 'D=41655 t=1645549347677971', 'Connection': 'close'}
2022-02-22 18:03:56,861:DEBUG:RESTInteractions,169:Sleeping 99 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:03:56 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '7dab92f4450089441529457bfe435a3d', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '5372.047 us', 'CMS-Server-Time': 'D=19146 t=1645549436840037', 'Connection': 'close'}
2022-02-22 18:05:35,915:DEBUG:RESTInteractions,169:Sleeping 122 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:05:35 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '797b22855a321bbd1347876e98135d91', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '6944.418 us', 'CMS-Server-Time': 'D=15355 t=1645549535900235', 'Connection': 'close'}
2022-02-22 18:07:38,054:DEBUG:RESTInteractions,169:Sleeping 145 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:07:38 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': 'd9061248e6d6f263f99e4b6f27c57b5e', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '4348.516 us', 'CMS-Server-Time': 'D=14096 t=1645549658038311', 'Connection': 'close'}
2022-02-22 18:10:03,205:DEBUG:RESTInteractions,169:Sleeping 166 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:10:03 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '78dc1e2ade2b94c6d1d24b9f3ae853b1', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '5981.922 us', 'CMS-Server-Time': 'D=18541 t=1645549803184421', 'Connection': 'close'}
2022-02-22 18:12:49,318:DEBUG:RESTInteractions,169:Sleeping 186 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:12:49 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': 'f6b028883e2a0c4615f3d5561a5210db', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '5544.901 us', 'CMS-Server-Time': 'D=17359 t=1645549969300218', 'Connection': 'close'}
2022-02-22 18:15:55,476:DEBUG:RESTInteractions,169:Sleeping 192 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:15:55 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': '6bffd89c2c0046250534c78506a68f22', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '4115.105 us', 'CMS-Server-Time': 'D=22826 t=1645550155451615', 'Connection': 'close'}
2022-02-22 18:19:03,563:INFO:MasterWorker,296:Received kill request. Setting STOP flag in the master process...
2022-02-22 18:19:07,525:DEBUG:RESTInteractions,169:Sleeping 220 seconds after HTTP error. Error details:  {'Date': 'Tue, 22 Feb 2022 17:19:07 GMT', 'Server': 'nginx/1.17.10', 'Content-Type': 'text/html;charset=utf-8', 'Content-Length': '745', 'X-Rest-Status': '403', 'X-Error-Http': '500', 'X-Error-Id': 'cf0b509b9377dced912a0054be919237', 'X-Error-Detail': 'Execution error', 'X-Rest-Time': '3882.885 us', 'CMS-Server-Time': 'D=13630 t=1645550347508886', 'Connection': 'close'}
```

### additional info

[1]

```plaintext
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA ENTER cms_analysis_reqmgr_w@cmsr CRABInterface.RESTBaseAPI.RESTBaseAPI (POST prod task) inuse=0 idle=2 - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA (previously uHEPNszRtCyn) connected, client: 19.11.0.0.0, server: 19.13.0.0.0, stmtcache: 50 - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA ping - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA check [select sysdate from dual] - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA connection established - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  DEBUG: authz operator {'dn': '/DC=ch/DC=cern/OU=computers/CN=tw/crab-prod-tw01.cern.ch', 'method': 'X509Cert', 'login': 'service@crab-prod-tw01.cern.ch', 'name': 'CRAB prod TW-01 Service', 'roles': {'operator': {'site': set(), 'group': {'crab3'}}}} to access resources ['220222_143422:jmsilva_crab_SPRACE-BR-tests-1645540093-Analysis_FileBased-3'] - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA disconnecting - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  SERVER DATABASE ERROR 500/403 Execution error binascii.Error 78dc1e2ade2b94c6d1d24b9f3ae853b1 [instance: prod] (Incorrect padding); last statement: None; binds: None, None; offset: None - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]    Traceback (most recent call last): - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]      File "/data/srv/HG2202e-b2671ebb11b2d65381339dff01177eb8/sw.crab_python3/slc7_amd64_gcc630/cms/crabserver/py3.220218-128bd804c35d83c081c2a5e59441f52b/lib/python3.8/site-packages/WMCore/REST/Server.py", line 1774, in dbapi_wrapper - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]        return handler(*xargs, **xkwargs) - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]      File "/data/srv/HG2202e-b2671ebb11b2d65381339dff01177eb8/sw.crab_python3/slc7_amd64_gcc630/cms/crabserver/py3.220218-128bd804c35d83c081c2a5e59441f52b/lib/python3.8/site-packages/CRABInterface/RESTTask.py", line 204, in post - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]        return getattr(RESTTask, subresource)(self, **kwargs) - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]      File "/data/srv/HG2202e-b2671ebb11b2d65381339dff01177eb8/sw.crab_python3/slc7_amd64_gcc630/cms/crabserver/py3.220218-128bd804c35d83c081c2a5e59441f52b/lib/python3.8/site-packages/CRABInterface/RESTTask.py", line 222, in addwarning - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]        warning = decodeBytesToUnicode(b64decode(kwargs['warning'])) - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]      File "/data/srv/HG2202e-b2671ebb11b2d65381339dff01177eb8/sw.crab_python3/slc7_amd64_gcc630/external/python3/3.8.2-comp/lib/python3.8/base64.py", line 87, in b64decode - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]        return binascii.a2b_base64(s) - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]    binascii.Error: Incorrect padding - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03] crabserver-879f66d87-g4kp8 137.138.120.6 "POST /crabserver/prod/task HTTP/1.1" 500 Internal Server Error [data: 1090 in 745 out 6335 us ] [auth: OK "/DC=ch/DC=cern/OU=computers/CN=tw/crab-prod-tw01.cern.ch" "" ] [ref: "" "CRABTaskWorker/py3.220203" ] - Podname=crabserver-879f66d87-g4kp8
[22/Feb/2022:18:10:03]  RESTSQL:MNueprGZEyiA DISCONNECTED cms_analysis_reqmgr_w@cmsr timeout=300 inuse=0 idle=1 - Podname=crabserver-879f66d87-g4kp8
WARNING. NOT AUTHORIZED - Podname=crabserver-879f66d87-g4kp8
```

[4] 

```diff
[crab3@crab-prod-tw01 TaskWorker]$ diff -u /data/repos/CRABServer/src/python/TaskWorker/MasterWorker.py /data/srv/TaskManager/py3.220203-fb21e34d414e2cf55ccf861e305e8c2c/slc7_amd64_gcc630/cms/crabtaskworker/py3.220203-fb21e34d414e2cf55ccf861e305e8c2c/lib/python3.8/site-packages/TaskWorker/MasterWorker.py 
--- /data/repos/CRABServer/src/python/TaskWorker/MasterWorker.py        2022-02-22 18:52:35.258138164 +0100
+++ /data/srv/TaskManager/py3.220203-fb21e34d414e2cf55ccf861e305e8c2c/slc7_amd64_gcc630/cms/crabtaskworker/py3.220203-fb21e34d414e2cf55ccf861e305e8c2c/lib/python3.8/site-packages/TaskWorker/MasterWorker.py      2022-02-22 18:23:03.546797695 +0100
@@ -356,13 +356,15 @@
             else:
                 failstatus = 'FAILED'
             self.updateWork(taskname, command, failstatus)
-            warning = 'username %s banned in CRAB TaskWorker configuration' % task['tm_username']
-            configreq = {'subresource': 'addwarning', 'workflow': taskname, 'warning': warning}
-            try:
-                self.crabserver.post(api='task', data=urlencode(configreq))
-            except Exception as e:
-                self.logger.error("Error uploading warning: %s", str(e))
-                self.logger.warning("Cannot add a warning to REST interface. Warning message: %s", warning)
+            #SB START comment out warning
+            #warning = 'username %s banned in CRAB TaskWorker configuration' % task['tm_username']
+            #configreq = {'subresource': 'addwarning', 'workflow': taskname, 'warning': warning}
+            #try:
+            #    self.crabserver.post(api='task', data=urlencode(configreq))
+            #except Exception as e:
+            #    self.logger.error("Error uploading warning: %s", str(e))
+            #    self.logger.warning("Cannot add a warning to REST interface. Warning message: %s", warning)
+            #SB END
             return True
         return False
```

**Comment by belforte** [comment]
the call which was failing some from
https://github.com/dmwm/CRABServer/blob/82e032dcebd3f68b4f971d8349dd2b722b92e3cb/src/python/TaskWorker/MasterWorker.py#L359-L366
it is the same API used in other places to add the warnings which lead to #7090 and which in that case worked fine. So I fear that we are left with "in this particular case the exact content of the warning string was such to trigger the padding error". I.e. a problem in REST code. Most likely the fact that b64decode throws error if length of arg. in bytes is not a multiple of 4. See e.g.
https://stackoverflow.com/questions/6102311/python-b64decode-incorrect-padding
https://gist.github.com/perrygeo/ee7c65bb1541ff6ac770
https://stackoverflow.com/a/49459036

I had met the b64decode problem already [1] but had not made the connection with the fact that we use it also in CRAB :-(  Sorry

So hopefully we have a simple solution here !


[1]
https://twiki.cern.ch/twiki/bin/view/Sandbox/TokensTips

**Comment by belforte** [comment]
maybe better to rename the issue to something like "protect b6decode from malformed strings" ? Up to you @mapellidario . OTOH it is odd that we do not make a proper-length string when urlencoding on client side. So maybe I am off-track here :-(

**Comment by belforte** [comment]
I believe that the initial description is incorrect.
TW does not get stuck on HTTP Error 500. But it does not know if such an error is a glitch or not so it retries blindly several times. Eventually it will give up and go on. The problem here is that there were O(100) tasks to fail for the banned user, so every TW slave was in the same loop (note that logs/twlog.txt as reported in [3] at the top comvines messages from 6 slaves). System was working "as designed", but of course retry loops only make sense if CRABSERVER failure is transient, not in cases like this.

**Comment by mapellidario** [comment]
I was wondering, what would happen if we dont retry errors `4XX`, but only `5XX`? is it possible that a problem with crabserver makes it return a `4XX`? 

Or should we simply improve the logic of the code here to make sure that when we have more than 10 warnings, we simply drop the last and move on with our task processing, without retrying anything?

**Comment by belforte** [comment]
please see: https://github.com/dmwm/CRABServer/issues/7090#issuecomment-1048618854
let's keep this issue here about avoiding fatal error from `addwarning` API.

**Comment by mapellidario** [comment]
> I believe that the initial description is incorrect.

Please, feel free to change both the issue title and the issue description as you like!

**Comment by mapellidario** [comment]
I think this is pretty easy to solve

Here is what we should do when we upload a warning:

https://github.com/dmwm/CRABServer/blob/466bcba95472b1b256c087016ae37c572d9a5f38/src/python/TaskWorker/Actions/TaskAction.py#L58-L62


Here is what we do in the MasterWorker

https://github.com/dmwm/CRABServer/blob/466bcba95472b1b256c087016ae37c572d9a5f38/src/python/TaskWorker/MasterWorker.py#L359-L362

In MasterWorker we need to `base64`-encode the warning. My bad, when I changed how we pass the warnings around i did not check if we uploaded a warning from a different place than TaskAction.

I will open a PR and test this idea in `preprod`.
